### PR TITLE
Fish 9690 filtering incorrect characters from header rfc 9110

### PR DESCRIPTION
**Pull request by breakponchito**
This is a PR to include new properties on the grizzly implementation to enable Header Name and Header Value validation against invalid characters mentioned on the RFC-9110, here the reference of the reported issue: [Glassfish issue](https://github.com/eclipse-ee4j/glassfish/issues/25128 ) and here the reserved CVE [CVE-2024-45687](https://www.cve.org/CVERecord?id=CVE-2024-45687)

To test this set the dependency version on the payara server: 4.0.2.payara-p2 build and run
deploy the following reproducer: 
[ReproducerJDK11.zip](https://github.com/user-attachments/files/17711134/ReproducerJDK11.zip)

after starting the server set the following properties on the Payara configuration. You can do it by console ui, command line or directly on the domain.xml file:

- -Dorg.glassfish.grizzly.http.STRICT_HEADER_NAME_VALIDATION_RFC_9110=true
- -Dorg.glassfish.grizzly.http.STRICT_HEADER_VALUE_VALIDATION_RFC_9110=true

image

image

`
asadmin create-jvm-options --target=server-config "-Dorg.glassfish.grizzly.http.STRICT_HEADER_NAME_VALIDATION_RFC_9110\=true"
`

`
asadmin create-jvm-options --target=server-config "-Dorg.glassfish.grizzly.http.STRICT_HEADER_VALUE_VALIDATION_RFC_9110\=true"
`

after adding the properties you need to restart the server

by default on the grizzly side those properties are set as false

deploy the reproducer and make curl calls

Header Name Validation tests:

-  curl -i -v -H $'X-MyHeader\n: hello' http://localhost:8080/PayaraServletProject-1.0-SNAPSHOT/PayaraServletProject
- curl -i -v -H $'X-MyHeader\r: hello' http://localhost:8080/PayaraServletProject-1.0-SNAPSHOT/PayaraServletProject

the result of any of those test should be error 400:
image

this will return status 200  but the header will be discarted because of the \0 character
- curl -i -v -H $'X-MyHeader\0: hello' http://localhost:8080/PayaraServletProject-1.0-SNAPSHOT/PayaraServletProject

image

Header Value Validations tests:
- curl -i -v -H $'X-MyHeader: he\nllo' http://localhost:8080/PayaraServletProject-1.0-SNAPSHOT/PayaraServletProject
image
- curl -i -v -H $'X-MyHeader: he\rllo' http://localhost:8080/PayaraServletProject-1.0-SNAPSHOT/PayaraServletProject 
image
- curl -i -v -H $'X-MyHeader: he\0llo' http://localhost:8080/PayaraServletProject-1.0-SNAPSHOT/PayaraServletProject
image

In the case of header validation we can't be very strict on the validation that is why the \n is permitted an is evaluated as a new line on the content for the header value and the \0 is immediately not processed from curl because by default we can't consider that character on the request headers and the \r character can be included as a character on the header values and this will be rejected

### CODE REVIEW

**Review by Pandrex247 — CHANGES_REQUESTED** [review]
Please rework to be semantically compatible with a minor version update (no breaking API changes)

**Comment by breakponchito** [comment]
> Please rework to be semantically compatible with a minor version update (no breaking API changes)

already fix the issue for the semantic versioning

File: modules/http/src/main/java/org/glassfish/grizzly/http/HttpCodecFilter.java [hunk 1]
```diff
@@ -710,6 +719,10 @@ protected boolean parseHeaderFromBytes(final HttpHeader httpHeader, final MimeHe
                 if (!parseHeaderName(httpHeader, mimeHeaders, parsingState, input, end)) {
                     return false;
                 }
+                
+                if(parsingState.subState == 0 && parsingState.start== -1) {
```
Comment (Pandrex247) [inline modules/http/src/main/java/org/glassfish/grizzly/http/HttpCodecFilter.java]:
```suggestion
                if (parsingState.subState == 0 && parsingState.start== -1) {
```

File: modules/http/src/main/java/org/glassfish/grizzly/http/HttpCodecFilter.java [hunk 2]
```diff
@@ -963,8 +1010,12 @@ protected boolean parseHeaderFromBuffer(final HttpHeader httpHeader, final MimeH
                 parsingState.subState++;
             }
             case 1: { // parse header name
-                if (!parseHeaderName(httpHeader, mimeHeaders, parsingState, input)) {
+                if(!parseHeaderName(httpHeader, mimeHeaders, parsingState, input)){
```
Comment (Pandrex247) [inline modules/http/src/main/java/org/glassfish/grizzly/http/HttpCodecFilter.java:1013]:
```suggestion
                if (!parseHeaderName(httpHeader, mimeHeaders, parsingState, input)){
```

**Review by Pandrex247 — DISMISSED** [review]
This is changing an existing Grizzly 4.0.0 patched branch to a 4.0.2 - this should be getting made to a new 4.0.2 maintenance branch, not updating an existing one

File: modules/http/src/main/java/org/glassfish/grizzly/http/HttpCodecFilter.java (hunk 1 above)
Comment (breakponchito) [reply modules/http/src/main/java/org/glassfish/grizzly/http/HttpCodecFilter.java]: already fix

File: modules/http/src/main/java/org/glassfish/grizzly/http/HttpCodecFilter.java (hunk 2 above)
Comment (breakponchito) [reply modules/http/src/main/java/org/glassfish/grizzly/http/HttpCodecFilter.java:1013]: already fixed

**Review by Pandrex247 — CHANGES_REQUESTED** [review]
Needs rebasing onto 4.0.0 without all of the 4.0.2 commits

**Pull request closed without merging**